ecosystem/sep-0031.md

```diff
@@ -100,11 +126,12 @@ All endpoints respond with content type:
 1. The Sending Anchor makes a `SEP-12 GET /customer` request for the Sending and Receiving Clients if required.
 1. The Sending Anchor makes a `SEP-12 PUT /customer` request for the Sending and Receiving Clients if required.
   1. The Receiving Anchor must validate the KYC data provided and reject the request with useful error messages if invalid.
+1. The Sending Anchor may request a quote using the `SEP-38 GET /price` or `SEP-38 POST /quote` endpoints.
 1. The Sending Anchor makes a `POST /transactions` request.
-  1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid.
+  1. The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide a `quote_id` instead of asset and amount information if [SEP-38](sep-0038.md) is supported.
```

Comment:
I don't feel strongly either way. If we made asset & amount mandatory in all cases, the other optional parameters would be `quote_id` and `destination_asset`, which would be mutually exclusive. I don't like the idea of also requiring `destination_asset` in all requests since requiring a new field is a breaking change for existing implementations.

Reply:
On second thought, requiring the user to repeat the fields that are already part of the quote goes against the [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) principle. 

@leighmcculloch, @jacekn: if you both agree with that I'll approve this PR and update #973 to use the same strategy [here](https://github.com/stellar/stellar-protocol/pull/973/#discussion_r674033709).

Reply:
It may be simpler, I think, to keep the API stable regardless if a quote is used, and to have the quote ID be an optional parameter. DRY isn't a truth, just a tool, and doesn't necessarily result in the simplest to understand APIs. In this specific case I do not feel strongly though and defer to you both.

Something that some APIs do to separate these changing input parameters from other parameters is to wrap them in a subfield and introduce a concept to wrap up those changing inputs. A good example of this is Stripe's source/payment intent interface. We could do something similar by saying the changing fields are the ExchangeParams, and for taht field you can either provide RawExchangeParams or QuoteExchangeParams.

Reply:
I really like this idea of a subfield, especially because the SEP-31 endpoint accepts `application/json` in a POST request. This pattern doesn't work well with SEP-6's `GET /*-exchange` endpoints though.

Reply:
Even if we don't use a subfield, thinking about it more the subfield identifies a conceptual oddity with using mutual exclusion on destination_asset. For example:

<table>
<tr><th>non-quote</th><th>quote</th></tr>
<tr>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "destination_asset": "...",
  },
}
```

</td>
<td>

```json
{
  "asset": "...",
  "amount": "...",
  "subfield": {
    "quote_id": "...",
  },
}
```

</td>
</tr>
</table>

It's not intuitive to me that destination asset is inside the subfield when asset is outside. In the same way it's not intuitive that destination_asset is optional when asset isn't.

Reply:
Since we're not creating a `POST /transactions-exchange` endpoint and instead using the original `POST /transactions` endpoint we cannot require `destination_asset` on every request, otherwise that would be a breaking change for existing applications.

Personally I don't think having two mutually exclusive parameters is too confusing, especially if we have a subsection below the request parameters list explaining when to use each one.

Reply:
Updated language:

> The Receiving Anchor must ensure the asset, amount, transaction fields, and customers IDs are valid. The Sending Anchor may provide either, but not both, a `quote_id` or `destination_asset` attribute if [SEP-38](sep-0038.md) is supported.

Reply:
Actually, I just realized that `destination_asset` can be specified if `quote_id` is also specified. There is no need for mutual exclusion.

So the three combinations are:
- `asset_code`, `amount`
- `asset_code`, `amount`, `destination_asset`
- `asset_code`, `amount`, `destination_asset`, `quote_id`

And if `quote_id` is specified, the information must match that of the quote.

Thoughts?